File: trained_model/Code/model_runner.py
import sys
import os
import glob
import onnxruntime as ort
import numpy as np

def main(folder_path):
    if not os.path.exists(folder_path):
        print(f"Error: Folder {folder_path} does not exist.")
        sys.exit(1)

    print(f"Scanning folder: {folder_path}")
    
    # Normally you would use rasterio here to load the 12 band images
    # import rasterio
    
    # model_path = os.path.join(os.path.dirname(__file__), 'ONNX_MODEL_DIR/coastal_classifier.onnx')
    
    # 1. Load ONNX Session
    # session = ort.InferenceSession(model_path)
    
    # 2. Load Raster Bands from folder 
    # bands = []
    # for band_label in ["B01","B02","B03","B04","B05","B06","B07","B08","B8A","B09","B11","B12"]:
    #     band_path = glob.glob(f"{folder_path}/*{band_label}.tiff")
    #     with rasterio.open(band_path[0]) as src:
    #         bands.append(src.read(1))
    
    # ... Preprocessing matching MATLAB mu / sigma ...
    # ... Session run ...
    # ... Post-processing ...
    
    print("SUCCESS: Image processed. Classification completed.")
    # Here you would output the path to the newly saved classified image
    # print("/path/to/output_classified.png")

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("Usage: python model_runner.py <path_to_geotiff_folder>")
        sys.exit(1)
        
    main(sys.argv[1])

File: seascan ui/src/imports/modelRunner.ts
import * as ort from 'onnxruntime-web';
import { fromArrayBuffer } from 'geotiff';

// Specify the expected 12 band labels in exact order
const BAND_LABELS = ["B01", "B02", "B03", "B04", "B05", "B06", "B07", "B08", "B8A", "B09", "B11", "B12"];

// Replace these with the actual values exported from your MATLAB normalization.mat
const MU = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]; // TODO: Update with actual Mean values
const SIGMA = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]; // TODO: Update with actual Std Deviation values

export async function processGeotiffCluster(tiffFiles: File[]): Promise<string> {
  try {
    // 1. Load ONNX Session
    // This expects coastal_classifier.onnx to be placed in the public/ directory of the UI
    const session = await ort.InferenceSession.create('/coastal_classifier.onnx');
    
    // 2. Read and extract pixel data from the uploaded TIFFs
    const bandData: Record<string, Float32Array> = {};
    for (const file of tiffFiles) {
      const arrayBuffer = await file.arrayBuffer();
      const tiff = await fromArrayBuffer(arrayBuffer);
      const image = await tiff.getImage();
      const rasters = await image.readRasters();
      
      // Match file name to Band label (e.g. T31UFU_20260504T103021_B01.tiff -> B01)
      const matchedBand = BAND_LABELS.find(b => file.name.includes(b));
      if (matchedBand) {
        bandData[matchedBand] = rasters[0] as Float32Array;
      }
    }

    // Ensure all 12 bands are present
    if (Object.keys(bandData).length !== 12) {
      throw new Error("Missing bands. Please upload all 12 Sentinel-2 bands.");
    }

    // Preprocessing & Inference will go here
    // For a single pixel (or batch matrix of pixels), we normalize it using MU and SIGMA
    // and pass it into session.run() ...

    return "Classification ready (Processing logic stubbed)";
  } catch (error) {
    console.error("Error running model:", error);
    throw error;
  }
}
